tests/src/test/java/org/ballerina/test/MiIntegrationTest.java

```
@@ -0,0 +1,341 @@
+import org.testcontainers.containers.wait.strategy.WaitAllStrategy;
+import org.testcontainers.images.builder.ImageFromDockerfile;
+import org.testng.Assert;
+import org.testng.SkipException;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.time.Duration;
+
+/**
+ * Integration test for MI project using Testcontainers and REST Assured.
+ * This test:
+ * 1. Builds a Docker image with the MI project CAR file
+ * 2. Starts a container with the MI server
+ * 3. Tests the API endpoints using REST Assured
+ */
+public class MiIntegrationTest {
+
+    private static final String MI_VERSION = System.getProperty("miVersion", "4.5.0");
+    private static final String BASE_IMAGE = "wso2/wso2mi:" + MI_VERSION;
+    private static final int MI_HTTP_PORT = 8290;
+    private static final int MI_HTTPS_PORT = 8253;
+    private static final int MI_MANAGEMENT_PORT = 9164;
+    
+    private static boolean dockerAvailable = false;
+    private static boolean setupCompleted = false;
+    private GenericContainer<?> miContainer;
+    private String containerBaseUrl;
+    private HttpClient httpClient;
+
+    @BeforeClass
+    public void setUp() throws Exception {
+        // Check if Docker is available
+        try {
+            Process dockerCheck = new ProcessBuilder("docker", "ps").start();
+            int exitCode = dockerCheck.waitFor();
+            dockerAvailable = (exitCode == 0);
+        } catch (Exception e) {
+            dockerAvailable = false;
+        }
+        
+        if (!dockerAvailable) {
+            System.out.println("Docker is not available. Skipping integration tests.");
+            setupCompleted = false;
+            return; // Exit early, don't throw SkipException in @BeforeClass
```

Shall we throw an exception here? So the other test cases may be skipped. You don't need to keep the flag and check where methods

---

To fail at test initialization